pkg/services/ngalert/notifier/channels/slack.go

```diff
@@ -35,132 +35,110 @@ type SlackNotifier struct {
 	tmpl          *template.Template
 	images        ImageStore
 	webhookSender notifications.WebhookSender
-
-	URL            *url.URL
-	Username       string
-	IconEmoji      string
-	IconURL        string
-	Recipient      string
-	Text           string
-	Title          string
-	MentionUsers   []string
-	MentionGroups  []string
-	MentionChannel string
-	Token          string
+	settings      slackSettings
 }
 
-type SlackConfig struct {
-	*NotificationChannelConfig
-	URL            *url.URL
-	Username       string
-	IconEmoji      string
-	IconURL        string
-	Recipient      string
-	Text           string
-	Title          string
-	MentionUsers   []string
-	MentionGroups  []string
-	MentionChannel string
-	Token          string
+type slackSettings struct {
+	EndpointURL      string   `json:"endpointUrl,omitempty" yaml:"endpointUrl,omitempty"`
+	URL              string   `json:"url,omitempty" yaml:"url,omitempty"`
+	Token            string   `json:"token,omitempty" yaml:"token,omitempty"`
+	Recipient        string   `json:"recipient,omitempty" yaml:"recipient,omitempty"`
+	Text             string   `json:"text,omitempty" yaml:"text,omitempty"`
+	Title            string   `json:"title,omitempty" yaml:"title,omitempty"`
+	Username         string   `json:"username,omitempty" yaml:"username,omitempty"`
+	IconEmoji        string   `json:"icon_emoji,omitempty" yaml:"icon_emoji,omitempty"`
+	IconURL          string   `json:"icon_url,omitempty" yaml:"icon_url,omitempty"`
+	MentionChannel   string   `json:"mentionChannel,omitempty" yaml:"mentionChannel,omitempty"`
+	MentionUsersStr  string   `json:"mentionUsers,omitempty" yaml:"mentionUsers,omitempty"`
+	MentionUsers     []string `json:"-" yaml:"-"`
```

Comment:
I am not sure I like such approaches with intermediate strings, especially when Go provides very simple ways to implement custom unmarshallers. I think it will be much cleaner to do something like 

<details> <summary> custom unmarshaler </summary>

```go
type CommaSeparatedStrings []string

func (r *CommaSeparatedStrings) UnmarshalJSON(b []byte) error {
	var str string
	if err := json.Unmarshal(b, &str); err != nil {
		return err
	}
	if len(str) > 0 {
		res := CommaSeparatedStrings(SplitString(str))
		*r = res
	}
	return nil
}

func (r *CommaSeparatedStrings) MarshalJSON() ([]byte, error) {
	if r == nil {
		return nil, nil
	}
	str := strings.Join(*r, ",")
	return json.Marshal(str)
}

func (r *CommaSeparatedStrings) UnmarshalYAML(b []byte) error {
	var str string
	if err := yaml.Unmarshal(b, &str); err != nil {
		return err
	}
	if len(str) > 0 {
		res := CommaSeparatedStrings(SplitString(str))
		*r = res
	}
	return nil
}

func (r *CommaSeparatedStrings) MarshalYAML() (interface{}, error) {
	if r == nil {
		return nil, nil
	}
	str := strings.Join(*r, ",")
	return yaml.Marshal(str)
}
```

</details>

Reply:
This is a really nice solution. Thanks!